hazelcast/src/main/java/com/hazelcast/sql/impl/worker/QueryOperationWorker.java

```diff
@@ -0,0 +1,172 @@
+        try {
+            while (true) {
+                Object task = queue.take();
+
+                if (task == POISON) {
+                    break;
+                } else {
+                    assert task instanceof QueryOperationExecutable;
+
+                    execute((QueryOperationExecutable) task);
+                }
+            }
+        } catch (InterruptedException e) {
+            // No-op.
+        }
+    }
+
+    private void execute(QueryOperationExecutable task) {
+        QueryOperation operation = task.getLocalOperation();
+
+        if (operation == null) {
+            operation = deserialize(task.getRemoteOperation());
+        }
+
+        operationHandler.execute(operation);
+    }
+
+    /**
+     * Deserializes the packet into operation. If an exception happens, the query is cancelled.
+     *
+     * @param packet Packet packet.
+     * @return Query operation.
+     */
+    private QueryOperation deserialize(Packet packet) {
+        try {
+            return ss.toObject(packet);
+        } catch (Exception e) {
+            if (e.getCause() instanceof QueryOperationDeserializationException) {
+                QueryOperationDeserializationException error = (QueryOperationDeserializationException) e.getCause();
+
+                // We assume that only ID aware operations may hold user data. Other operations contain only HZ classes and
+                // we should never see deserialization errors for them.
+                sendDeserializationError(error);
+            } else {
+                // It is not easy to decide how to handle an arbitrary exception. We do not have caller coordinates, so
+                // we do not know how to notify it. We also cannot panic (i.e. kill local member), because it would be a
+                // security threat. So the only sensible solution is to log the error.
+                logger.severe("Failed to deserialize query operation received from " + packet.getConn().getEndPoint()
+                    + " (will be ignored)", e);
+            }
+        }
+
+        return null;
+    }
+
+    private void sendDeserializationError(QueryOperationDeserializationException e) {
+        QueryId queryId = e.getQueryId();
+        UUID callerId = e.getCallerId();
+
+        HazelcastSqlException error = HazelcastSqlException.error("Failed to deserialize " + e.getOperationClassName()
+            + " received from " + callerId + ": " + e.getMessage(), e);
+
+        QueryCancelOperation cancelOperation =
+            new QueryCancelOperation(queryId, error.getCode(), error.getMessage(), localMemberId);
+
+        try {
+            operationHandler.submit(queryId.getMemberId(), cancelOperation);
+        } catch (Exception ignore) {
+            // This should never happen, since we do not transmit user objects.
+        }
+    }
+
+    /**
+     * For testing only.
+     */
+    boolean isThreadTerminated() {
+        return thread.getState() == Thread.State.TERMINATED;
```

Comment:
`!thread.isAlive()` ?

Reply:
It doesn't matter much because the method is used for testing only.